Source/Core/DolphinQt2/Config/ARCodeWidget.cpp

```diff
@@ -0,0 +1,175 @@
+
+  auto* button_layout = new QHBoxLayout;
+
+  button_layout->addWidget(m_code_add);
+  button_layout->addWidget(m_code_edit);
+  button_layout->addWidget(m_code_remove);
+
+  QVBoxLayout* layout = new QVBoxLayout;
+
+  layout->addWidget(m_warning);
+  layout->addWidget(m_code_list);
+  layout->addLayout(button_layout);
+
+  setLayout(layout);
+}
+
+void ARCodeWidget::ConnectWidgets()
+{
+  connect(m_warning, &CheatWarningWidget::OpenCheatEnableSettings, this,
+          &ARCodeWidget::OpenGeneralSettings);
+  connect(m_code_list, &QListWidget::itemChanged, this, &ARCodeWidget::OnItemChanged);
+  connect(m_code_list, &QListWidget::itemSelectionChanged, this, &ARCodeWidget::OnSelectionChanged);
+
+  connect(m_code_add, &QPushButton::pressed, this, &ARCodeWidget::OnCodeAddPressed);
+  connect(m_code_edit, &QPushButton::pressed, this, &ARCodeWidget::OnCodeEditPressed);
+  connect(m_code_remove, &QPushButton::pressed, this, &ARCodeWidget::OnCodeRemovePressed);
+}
+
+void ARCodeWidget::OnItemChanged(QListWidgetItem* item)
+{
+  m_ar_codes[m_code_list->row(item)].active = (item->checkState() == Qt::Checked);
+  SaveCodes();
+}
+
+void ARCodeWidget::OnSelectionChanged()
+{
+  auto items = m_code_list->selectedItems();
+
+  if (items.empty())
+    return;
+
+  auto selected = items[0];
+
+  bool user_defined = m_ar_codes[m_code_list->row(selected)].user_defined;
+
+  m_code_remove->setEnabled(user_defined);
+  m_code_edit->setText(user_defined ? tr("Edit Code...") : tr("Clone and Edit Code..."));
+}
+
+void ARCodeWidget::UpdateList()
+{
+  m_code_list->clear();
+
+  for (auto& ar : m_ar_codes)
+  {
+    auto* item = new QListWidgetItem(QString::fromStdString(ar.name)
+                                         .replace(QStringLiteral("&lt;"), QStringLiteral("<"))
+                                         .replace(QStringLiteral("&gt;"), QStringLiteral(">")));
+
+    item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsSelectable | Qt::ItemIsUserCheckable);
+    item->setCheckState(ar.active ? Qt::Checked : Qt::Unchecked);
+
+    m_code_list->addItem(item);
+  }
+}
+
+void ARCodeWidget::SaveCodes()
+{
+  IniFile game_ini_local;
+  game_ini_local.Load(File::GetUserPath(D_GAMESETTINGS_IDX) + m_game_id + ".ini");
+  ActionReplay::SaveCodes(&game_ini_local, m_ar_codes);
+
+  game_ini_local.Save(File::GetUserPath(D_GAMESETTINGS_IDX) + m_game_id + ".ini");
+}
+
+void ARCodeWidget::OnCodeAddPressed()
+{
+  ActionReplay::ARCode ar;
+
+  ARCodeEditor ed(ar);
+
+  ed.exec();
+
+  m_ar_codes.push_back(ar);
+
+  UpdateList();
+  SaveCodes();
+}
+
+void ARCodeWidget::OnCodeEditPressed()
+{
+  auto items = m_code_list->selectedItems();
+
+  if (items.empty())
+    return;
+
+  auto selected = items[0];
+
+  bool user_defined = m_ar_codes[m_code_list->row(selected)].user_defined;
+
+  ActionReplay::ARCode ar = m_ar_codes[m_code_list->row(selected)];
+
+  ARCodeEditor ed(user_defined ? m_ar_codes[m_code_list->row(selected)] : ar);
```

Comment:
related to the above comment (if the copy is not intended) `m_ar_codes[m_code_list->row(selected)]` and `ar` would be the same thing here. (if the copy is intended, disregard this).